[INFINITY-1592] Automated TLS provisioning

mhrabovcin opened the pull request:
WIP: https://jira.mesosphere.com/browse/INFINITY-1592

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/ServiceAccountIAMTokenProvider.java:
```diff
@@ -0,0 +1,154 @@
+
+        Response response = httpExecutor.execute(request);
+
+        JSONObject resposneData = new JSONObject(response.returnContent().asString());
```
Fix typo to `responseData`.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/OfferEvaluator.java:
```diff
@@ -271,6 +275,15 @@ private static void logOutcome(StringBuilder stringBuilder, EvaluationOutcome ou
                         new VolumeEvaluationStage(volumeSpec, taskName, Optional.empty(), Optional.empty()));
             }
 
+            // TODO(mh): Improve error handling
```
We definitely need it.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java:
```diff
@@ -0,0 +1,291 @@
+        try {
+            keyPairGenerator = KeyPairGenerator.getInstance("RSA");
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
```
Stacktraces should be logged as errors and the method should return an exception, right?

mhrabovcin replied:
I don't know what is the best way to handle this as I don't think there is a chance that this will throw an exception.

pires replied:
Best option  is to check with the Infinity team.

gabrielhartmann replied:
`logger.error("Failed to generate key pair.", e)`.  I'm just reading the comment.  Don't know if you want to rethrow an `IllegalStateException` or not here.  Probably not, just fail the evaluation stage, and let the system retry.

vishnu2kmohan commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java:
```diff
@@ -0,0 +1,291 @@
+
+        GeneralNames subAtlNames = new GeneralNames(
+                new GeneralName[]{
+                        new GeneralName(GeneralName.dNSName, "test.com"),
```
Is this intentional? Looks like `test.com` and `127.0.0.1` are accidentally copied from the test code further below?

mhrabovcin replied:
Indeed, this is a WIP test code. I'll create a proper service name generator.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/ServiceAccountIAMTokenProvider.java:
```diff
@@ -0,0 +1,154 @@
+    private RSAPrivateKey privateKey;
+    private Executor httpExecutor;
+
+    public ServiceAccountIAMTokenProvider(
```
I hope this functionality will be extracted to a different repo, e.g. `dcos-ca-client-java` and just be imported here as a dependency. So I'm fine with public visibility for now.

benclarkwood replied:
There's a builder, so there's no need (and there really shouldn't be) public constructors.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/ServiceAccountIAMTokenProvider.java:
```diff
@@ -0,0 +1,154 @@
+        this.httpExecutor = executor;
+    }
+
+    public ServiceAccountIAMTokenProvider(Builder builder) {
```
Same comment as above.

mhrabovcin replied:
Changed both constructors to private and left builder as a way to create `ServiceAccountIAMTokenProvider`.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/secrets/DefaultSecretsClient.java:
```diff
@@ -0,0 +1,110 @@
+     * @throws SecretsException
+     */
+    protected void handleResponseStatusLine(StatusLine statusLine, int okCode, String path) throws SecretsException {
+        if (statusLine.getStatusCode() == okCode) {
```
Good catch @benclarkwood.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java:
```diff
@@ -0,0 +1,291 @@
+
+public class TLSEvaluationStage implements OfferEvaluationStage {
+
+    private static final Logger LOGGER = (Logger) LoggerFactory.getLogger(TLSEvaluationStage.class);
```
If there's only one instance of this class, indeed no need for static logger. +1 for cast removal.

mhrabovcin replied:
I am going to keep it static because static methods are using the logger.

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java:
```diff
@@ -0,0 +1,46 @@
+ * getToken() method on underlying provider when token is about to expire.
+ *
+ * // TODO(mh): This could be also running in a separate thread in background.
+ * // TODO(mh): Does this needs to be threadsafe?
```
@benclarkwood you are right. My concern was that multiple objects would try to get access and possibly refresh authentication token that would lead to a race condition.

pires replied:
If you'll be refreshing the token in the background, do make sure that you have a mutex when setting the token value.

benclarkwood replied:
We might as well make this Threadsafe to start.

@mhrabovcin let me know if you want guidance on doing so.

pires replied:
So @benclarkwood now you know why I said it would need to become thread-safe.

nickbp replied:
Should be sufficient to just lock all of `getToken()`.

mhrabovcin replied:
I've added `synchronized` keyword. Is that sufficient?

nickbp replied:
I'd lean towards having a lock object, it's pretty common for people to not understand what `synchronized`'s semantics are. See `MemPersister` or `PersisterCache` for examples of explicit lock objects.

mhrabovcin replied:
Thanks for the example. I've replaced `synchronized` with the `ReentrantReadWriteLock`.

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/ServiceAccountIAMTokenProvider.java:
```diff
@@ -0,0 +1,154 @@
+                    .withExpiresAt(Date.from(Instant.now().plusSeconds(120)))
+                    .sign(getRSA256Algorithm());
+        } catch (InvalidKeySpecException e) {
+            e.printStackTrace();
```
Correct, there shouldn't happen as we control RSA256 algorithm construction. I've added some logging and the code is still swallowing exceptions.

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/CertificateAuthorityClient.java:
```diff
@@ -0,0 +1,15 @@
+ */
+public interface CertificateAuthorityClient {
+
+    public X509Certificate sign(byte[] csr) throws IOException, CertificateException;
```
It indeed yelled at me in very polite way :-)

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/http/HttpClientBuilder.java:
```diff
@@ -0,0 +1,111 @@
+
+    public HttpClientBuilder setLogger(Logger logger) {
+
+        this.addInterceptorLast(new HttpRequestInterceptor() {
```
Converted both instances to lambdas.

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/http/URLHelper.java:
```diff
@@ -0,0 +1,45 @@
+        try {
+            return new URL(url);
+        } catch (MalformedURLException e) {
+            e.printStackTrace();
```
This meant to be a common helper where developers can decide whether they trust that URL is correct and that exception won't be raised while appending a path to a URL.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java:
```diff
@@ -31,10 +32,10 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
```
Honestly, I think this is a bad practice and imports should be explicit.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/security/CertificateNamesGenerator.java:
```diff
@@ -0,0 +1,50 @@
+     */
+    public GeneralNames getSANs() {
+        // TODO(mh): Should we dynamically test whether the network is overlay? If the network isn't
+        // overlay there is no vipName available.
```
Why add the extra complexity? The certificate will still be valid whether overlay is there or not.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/security/CertificateNamesGenerator.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * A {@link CertificateNamesGenerator} creates relevant names for given service pod.
+ *
+ * // TODO(mh): What about IP per container feature??
```
The problem with that is that you can't know the IP before the container actually runs, right? I'd say that as a first pass, we should support DNS alone and make it very clear in documentation. An issue to revisit this should then be added to the backlog.

benclarkwood replied:
Yes, first pass should just be DNS (IIRC we discussed this in slack). Obviously not 100% ideal, but best we can presently do.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/secrets/NotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+/**
+ * An operation requires existing secret which wasn't found.
+ */
+public class NotFoundException extends SecretsException {
```
Rename to `SecretNotFoundException`?

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java:
```diff
@@ -0,0 +1,276 @@
+
+        // Share keys to the container
+        Optional<Protos.ExecutorInfo.Builder> executorBuilder = podInfoBuilder.getExecutorBuilder();
+        Protos.TaskInfo.Builder taskBuilder = podInfoBuilder.getTaskBuilder(taskName);
```
nit: no need to assign `taskBuilder` here since we'll only refer to it `if (executorBuilder.isPresent())`. Move it inside the if statement.

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java:
```diff
@@ -0,0 +1,276 @@
+                    .setMode(Protos.Volume.Mode.RO)
+                    .build());
+
+        }
```
Can't recall now if you're using Java 8 elsewhere, but this could be replaced with a lambda:
```java
        tlsSecrets.entrySet().forEach(tlsSecretEntry ->
                volumes.add(Protos.Volume.newBuilder()
                        .setSource(Protos.Volume.Source.newBuilder()
                                .setType(Protos.Volume.Source.Type.SECRET)
                                .setSecret(getReferenceSecret(tlsSecretEntry.getKey()))
                                .build())
                        .setContainerPath(tlsSecretEntry.getValue())
                        .setMode(Protos.Volume.Mode.RO)
                        .build())
        );
```

pires commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java:
```diff
@@ -0,0 +1,276 @@
+        LOGGER.info(String.format("Creating new secret: %s", getSecretPath(podInfoBuilder, "certificate-pem")));
+
+        secretsClient.create(
+                getSecretPath(podInfoBuilder, "certificate-crt"),
```
`certificate-crt` or `certificate-pem` as in the log instruction above?

pires replied:
Also, there's duplicated `getSecretPath` calls for the same name. I suggest extracting those to variables.

mhrabovcin replied:
Moved to single `SecretNameGenerator` that can construct secret and volume mount paths for all types of artifacts.

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/StaticTokenProvider.java:
```diff
@@ -0,0 +1,20 @@
+ * Static Token Provider always returns single pre-configured auth token. It can be used for testing and for known
+ * life of configured token.
+ */
+public class StaticTokenProvider implements TokenProvider {
```
how about `ConstantTokenProvider`. static could be misinterpreted to mean global

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java:
```diff
@@ -0,0 +1,49 @@
+
+    public CachedTokenProvider(TokenProvider provider, int triggerRefreshBeforeSeconds) {
+        this.provider = provider;
+        this.triggerRefreshBeforeSeconds = triggerRefreshBeforeSeconds;
```
could just call it `ttl`.

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/http/DcosHttpClientBuilder.java:
```diff
@@ -0,0 +1,107 @@
+
+            public void checkServerTrusted(X509Certificate[] certs, String authType) {
+            }
+
```
lots of extra newlines like this throughout this file

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/http/DcosHttpClientBuilder.java:
```diff
@@ -0,0 +1,107 @@
+                request.addHeader("Authorization", String.format("token=%s", provider.getToken().getValue())));
+
+        return this;
+
```
newlines

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/http/DcosHttpClientBuilder.java:
```diff
@@ -0,0 +1,107 @@
+     * @return
+     */
+    public DcosHttpClientBuilder setTokenProvider(TokenProvider provider) {
+
```
extra

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/secrets/AlreadyExistsException.java:
```diff
@@ -0,0 +1,7 @@
+/**
+ * An operation is attempting to create a secret that already exists.
+ */
+public class AlreadyExistsException extends SecretsException {
```
IMO just have a single `SecretsException` class with a `reason` field. Makes it easier to log the cause of the exception without involving reflection. See `PersisterException` for an example of using a `reason` field. Also, the `store` and `path` fields being passed in the 401 case could just always be passed to `SecretsException` as they give more context as to the originating request.

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/secrets/DefaultSecretsClient.java:
```diff
@@ -0,0 +1,137 @@
+        return URLHelper.addPathUnchecked(baseUrl, path);
+    }
+
+
```
extra newline

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/CertificateAuthorityClient.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Create a new certificate from CSR by contacting certificate authority.
+     *
+     * @param csr
```
Could either:
- Remove doc stubs
- Fill doc stubs
(here and elsewhere)

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/CertificateAuthorityClient.java:
```diff
@@ -0,0 +1,36 @@
+     * @throws IOException
+     * @throws CertificateException
+     */
+    X509Certificate sign(byte[] csr) throws IOException, CertificateException, CAException;
```
Are any callers expected to behave differently depending on which of these exceptions is thrown? E.g. looks like CAException is only being checked in tests. IMO this could just be `throws Exception` or `throws IOException` (with CertificateException/CAException extending IOException in the latter case).

mhrabovcin replied:
I've changed the method signature to throw an `Exception`.

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class CachedTokenProvider implements TokenProvider {
+
+    private TokenProvider provider;
```
mark `provider` and `triggerRefreshBeforeSeconds` (or `ttl`, see below) as `final`

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/ServiceAccountIAMTokenProvider.java:
```diff
@@ -0,0 +1,158 @@
+ */
+public class ServiceAccountIAMTokenProvider implements TokenProvider {
+
+    private URL iamUrl;
```
can these members be marked `final`?

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/StaticTokenProvider.java:
```diff
@@ -0,0 +1,20 @@
+ */
+public class StaticTokenProvider implements TokenProvider {
+
+    private Token token;
```
`final`

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/Token.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Token represents a JWT authentication token issued by DC/OS Bouncer IAM service.
+ */
+public class Token {
```
This doesn't seem to add much value over just passing around the `DecodedJWT`. Any reason to not just use that?

mhrabovcin replied:
I was only trying to make package boundary and not leak underlying library into dcos auth implementation. I can easily remove `Token` class and use `DecodedJWT` instead.

nickbp replied:
I'd lean towards just exposing it, if we switch libraries it shouldn't be a big deal to swap it out

mhrabovcin replied:
👍 I've removed `Token` class and changed the code to use `DecodedJWT`.

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/secrets/ForbiddenException.java:
```diff
@@ -0,0 +1,21 @@
+    /**
+     * @return A string representing DC/OS permission required to operate on given secret.
+     */
+    public String getMissingPermission() {
```
doesn't look like anything is actually using/logging this?

benclarkwood commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/secrets/Secret.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Secret class represents an HTTP API model for creating and updating a secret.
+ */
+public class Secret {
```
How / where is this getting serialized out? You could use Jackson notations to simplify things if it's going to JSON.

mhrabovcin replied:
I was originally using the `JSONObject` but now I've switched to the `ObjectMapper`.

benclarkwood commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/security/CertificateNamesGenerator.java:
```diff
@@ -0,0 +1,64 @@
+     * @return
+     */
+    public X500Name getSubject() {
+       // TODO(mh): Should we somehow improve certificate name here? Right now it reflects only service
```
What are valid values for the common name? Any reason to not shove everything in?

mhrabovcin replied:
So there is a single Common Name and then there are Subject Alternative Name values. I am going to add `autoip` name to the `CN` based on https://knowledge.rapidssl.com/support/ssl-certificate-support/index?page=content&actp=CROSSLINK&id=SO7239.

>The Common Name (CN), also known as the Fully Qualified Domain Name (FQDN), is the characteristic value within a Distinguished Name. Typically, it is composed of Host Domain Name and looks like, "www.symantec.com" or "symantec.com". The Common Name field is often misinterpreted and is filled out incorrectly. Do not use your organization's actual name as your common name.

Based on https://docs.mesosphere.com/1.9/networking/dns-overview/ it seems that `autoip` should support accessing service on and off virtual networks.

Example name: `kafka-0-broker.kafka.autoip.dcos.thisdcos.directory`

mhrabovcin commented on sdk/bootstrap/main.go:
```diff
@@ -305,6 +314,51 @@ func installDCOSCertIntoJRE() {
 	log.Println("Successfully installed the certificate.")
 }
 
+func decodeKeystores() error {
```
There are currently no tests for bootstrap utility. I can add `decode` tests and add test evaluation to the build pipeline.

There is also gradle plugin that we could use to build this package instead of custom bash script -  https://github.com/echocat/gradle-golang-plugin .

benclarkwood replied:
1. Please add a test, but no need to wire it up yet
2. Does that plugin support the fact that we're outside of GOPATH? (e.g. that we shoehorn ourselves into it)

mhrabovcin replied:
Completed: 1, 3

I am picking from go gradle documentation:

> The gradle-golang plugin is is designed to give you the power of a professional build tool (in this case Gradle) to build, test (and other great stuff) with your go code. And this without pain of think about the correct installation go sdk, set the right environment variables, download dependencies, ...

> No pre installed go toolchain is required - only a working JDK (in minimum version 1.7).
> Choice to build in the default system GOPATH or in a temporary one to do not influence your build result

It looks like the plugin could be a replacement for our custom `golang` build scripts and it would be easier setting up dev environment as no `go` installation is required.

mhrabovcin replied:
https://jira.mesosphere.com/browse/INFINITY-2091

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java:
```diff
@@ -0,0 +1,68 @@
+ * CachedTokenProvider retrieves token from underlying provider and caches the value. It automatically triggers
+ * getToken() method on underlying provider when token is about to expire.
+ */
+public class CachedTokenProvider implements TokenProvider {
```
I am happy to use `Guava`. We don't use it yet, is it OK to add it as scheduler dependency?

benclarkwood replied:
Ah, didn't realize we weren't using it yet.

I'd stick with what you have in that case.

benclarkwood commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultTransportEncryptionSpec.java:
```diff
@@ -0,0 +1,84 @@
+public class DefaultTransportEncryptionSpec implements TransportEncryptionSpec {
+
+    @Valid
+    private String name;
```
To be in line with other required strings this should be:
```
@NotNull(message = "Service name cannot be empty")
@Size(min = 1, message = "Service name cannot be empty")
```

benclarkwood replied:
(The `@valid` simply tells the validation to ensure that `name` is validated without actually specifying what the validators are)

benclarkwood commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/security/TLSArtifactsPersister.java:
```diff
@@ -0,0 +1,140 @@
+     * use {@link TLSArtifactsPersister ::cleanUpSecrets} call.
+     * @param tlsArtifacts
+     */
+    public void persist(
```
Why does this persist _everything_ when the user may only have specified TLS and not KEYSTORE?

mhrabovcin commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+
+Every item in the `transport-encryption` configuration section must have unique `name`. The `type` represents a format in which the private key and certificate will be delivered into the task sandbox. Different items will result in a unique private key and certificate.
+
+The TLS artifacts within single task will be shared across all instances of the pod task. Different tasks can request TLS artficats with same name but each task will get unique private key and certificate and they won't get shared across different tasks.
```
I was under the impression that only the pod has an "identity" and you can tell to which pod replica task belongs, i.e. `pod-[index]-task`. I also thought you can run multiple copies of a single task.

mhrabovcin commented on sdk/bootstrap/secret_decoder.go:
```diff
@@ -0,0 +1,74 @@
+	}
+
+	// Decode each file
+	for _, path := range matches {
```
Added.

mhrabovcin commented on frameworks/kafka/src/main/dist/server.properties.mustache:
```diff
@@ -27,9 +27,31 @@ broker.id={{POD_INSTANCE_INDEX}}
 
-# Hostname and port the broker will advertise to producers and consumers. If not set, 
+{{#KAFKA_ENABLE_TLS}}
+############################# TLS Settings #############################
```
1. Technically its an easy solution but I don't see what additional security guarantees we'd get by creating a random keystore secret per task. Password would get injected to a container same way as JKS files. Also Java apps usually read a keystore password from a configuration file, which gets rendered by the `bootstrap` utility and gets written to a sandbox. If an attacker would get access to sanbox directory of any container (even not active) it would be possible to read the password and thus read a private key.

mhrabovcin commented on frameworks/helloworld/tests/test_tls.py:
```diff
@@ -0,0 +1,363 @@
+    # by SAN in certificate.
+    assert 'HTTP/1.1 200 OK' in output
+    assert 'CAfile: secure-tls-pod.ca' in output
+    assert 'host "keystore-https.hello-world.l4lb.thisdcos.directory" matched cert\'s "*.hello-world.l4lb.thisdcos.directory"' in output
```
I've added explicit VIP to SANs. There won't be any wildcards at all.

jgehrcke commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+
+The SDK provides automated way of provisioning X.509 certificates and private keys referered in this documentation also as TLS artifacts. TLS artifacts can be used to expose service over encrypted TLS connection.
+
+One or more TLS artifacts can be requrested by adding `transport-encryption` key to the `task` YAML definition.
```
typo: requrested

jgehrcke commented on docs/pages/developer-guide.md:
````diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+                    port: 80
+```
+
+Every item in the `transport-encryption` configuration section must have unique `name`. The `type` represents a format in which the private key and certificate will be delivered into the task sandbox. Different items will result in a unique private key and certificate.
````
1st sentence: [...] must have *a* unique name.

2nd sentence: The `type` field defines the serialization format with which the private key and certificate files will be written into the task sandbox.

3rd sentence: Each item will result in a unique private key and the corresponding certificate.

jgehrcke commented on docs/pages/developer-guide.md:
````diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+               tls.crt
+               tls.ca
+               ...
+```
````
Let's add a paragraph about those files. I speculate a bit with the following:

Here, the file `tls.crt` contains an end-entity certificate in the OpenSSL PEM format (if applicable, this file also includes corresponding intermediate CA certificates). The `tls.key` contains the private key corresponding to the end-entity certificate, in the PKCS#8 PEM format. The file `tls.ca` contains the DC/OS CA bundle in the OpenSSL PEM format.

jgehrcke replied:
Looking at it this way, the type "tls" should probably have been called "PEM" or "OpenSSL" :-).

mhrabovcin replied:
Agreed - https://docs.google.com/document/d/1eOcR3VoI2tfk7wjTmjkjOP7jF521J34uOdnm8bMKQaY/edit?disco=AAAABM5I3JI I've raised this concern earlier.

mhrabovcin replied:
> The file tls.ca contains the DC/OS CA bundle in the OpenSSL PEM format.

This is actually only root CA certificate. We're using CA's `bundle` API call that adds whole certificate chain.

jgehrcke commented on docs/pages/developer-guide.md:
````diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+hello-world/hello-0-server__tls__truststore
+```
+
+### TLS artifacts secrets lifetime
````
Lifetime of secrets containing the TLS artifacts

jgehrcke commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+
+## Private key and certificate details
+
+A scheduler provisions *private key* and `X.509` *certificate* and exposed them to task in various encodings.
```
- "exposes" instead of "exposed"
- "serialization formats" instead of "encodings" (that's the term that is typically used in libraries that want to abstract across various serialization formats)

jgehrcke commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+
+### X.509 certificate
+
+A `X.509` certificate for the private key is generated by [DC/OS certificate authority](https://docs.mesosphere.com/1.9/networking/tls-ssl/ca-api/) - `DC/OS CA`. A certificate is signed by DC/OS signing certificate. The signing certificate can be either self-signed root CA certificate created during the cluster installation or user provided signing CA certificate.
```
An X.509 end-entity certificate corresponding to the private key is generated with the help of the [DC/OS CA](link) by sending it a certificate signing request (CSR, built from the public key). The returned end-entity certificate is signed with the private key corresponding to the signing CA certificate that the DC/OS CA is configured with. That signing CA certificate can either be a root CA certificate automatically created during DC/OS cluster installation or a user-provided (custom) CA certificate.

jgehrcke commented on docs/pages/developer-guide.md:
````diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+A certificate has following subject information:
+
+```
+CN=[pod-index-task].[service-name].autoip.dcos.thisdcos.directory
````
How do we deal with the 64 char limitation for this? Do we auto-truncate or does cert creation then fail?

(one of the many RFCs requires the CN to not go beyond 64 chars)

mhrabovcin replied:
I wasn't aware of this limitation. This is something I have to address in the code and I am thinking now about using only `[pod-index-task].[service-name]`.

benclarkwood replied:
I think that's a reasonable compromise. From a quick internet survey, OpenSSL enforces this, but other things do not (since it conflicts with the limits of DNS). But, AFAIK we merely need our CNs to be consistent, yes?

jgehrcke replied:
> I am thinking now about using only [pod-index-task].[service-name].

What if this gets longer than 64 chars? The code that generates the subject string must handle the limit (sanitize input), otherwise I'd consider this to be a bug.

> OpenSSL enforces this, but other things do not (since it conflicts with the limits of DNS)

To be clear, [RFC 5280](https://www.ietf.org/rfc/rfc5280.txt) page 123 defines `ub-common-name INTEGER ::= 64` as an upper bound.

A certificate that does not adhere to this is an invalid X.509 certificate. I expect parts of the software stack here to simply throw an error if this string ever gets too long. Maybe that's not the case, if e.g. cfssl happily issues such certificates. But then.. we're maybe just "lucky".

I think the subject CNs constructed here should not have any importance for the overall system behavior. They are just a gimmick for having a nice human-readable representation of what this certificate is used for.

I know, there are points of view such as https://groups.google.com/forum/#!topic/mailing.openssl.users/SKTRhymXjXc -- but really, DNS Names should these days not be an argument anymore for not respecting this limit. We use Subject Alternative Names for hostname verification, also specificed by https://www.ietf.org/rfc/rfc5280.txt.

jgehrcke commented on docs/pages/developer-guide.md:
````diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+C=US
+```
+
+Additionaly each certificate gets `Subject Alternative Names`:
````
Additionally, the following X.509 Subject Alternative Names are encoded into the certificate:

jgehrcke commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+
+#### PEM
+
+Standard [PEM encoded](https://en.wikipedia.org/wiki/Privacy-enhanced_Electronic_Mail) certificate, private key and root CA certificate. Based on a file extension name it is possible to tell which artifact is in the file.
```
In the example above I have seen the file extension `.tls`.

jgehrcke replied:
For the record, because this is really important for us to know (internally, at least): there is no "PEM standard" that applies here. There is the non-standard "OpenSSL PEM format" (not covered by any RFC, just common enough to be a pseudo standard best described with the name OpenSSL PEM format).

jgehrcke replied:
With "root CA certificate" I think we mean DC/OS CA bundle. Correct?

mhrabovcin replied:
I've removed word **standard** to avoid possible confusion. The CA file contains only the corresponding root CA certificate and not the DC/OS CA bundle.

jgehrcke commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+
+A **.key** file contains PEM encoded private key.
+
+A file with **.crt** extension contains root CA certificate without certificate chain.
```
Is that, more generally, "the DC/OS CA bundle"?

(I propagate this term so that in the future we can have more than one trust anchor in there :-))

jgehrcke commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +936,160 @@ In the example above, a server task can be accessed through the address:
+
+## Installation requirements
+
+To enable TLS support a `Mesosphere DC/OS Enterprise` cluster must be installed in `permissive` or `strict` security mode.
```
To enable the TLS support described above ....

sascala replied:
comma after "above"

jgehrcke commented:
Hey ho hey; I have left a few comments in the documentation commit. This was not yet a thorough review, but what I saw look generally very good! Awesome stuff. :rocket:

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/dcos/auth/CachedTokenProviderTest.java:
```diff
@@ -0,0 +1,71 @@
+        verify(mockProvider, times(2)).getToken();
+    }
+
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/dcos/auth/ServiceAccountIAMTokenProviderTest.java:
```diff
@@ -0,0 +1,77 @@
+
+    }
+
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/dcos/auth/ServiceAccountIAMTokenProviderTest.java:
```diff
@@ -0,0 +1,77 @@
+        return new KeyPair(publicKey, privateKey);
+    }
+
+    // Following tests expects DC/OS to be running on 172.17.0.2 IP address and its not ran by default.
```
The unit tests don't use the network. Use mocks to simulate DC/OS behavior.

mhrabovcin replied:
Got it. I was using these methods to test responses against running cluster. Right now they are marked to be ignored by test framework using the `@Ignore` annotation but I am happy to remove them as we have mocked responses and proper unit tests in place.

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/dcos/http/URLHelperTest.java:
```diff
@@ -0,0 +1,18 @@
+        Assert.assertEquals(path, "/ca/api/v2/sign");
+    }
+
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStageTest.java:
```diff
@@ -0,0 +1,352 @@
+                .findAny();
+    }
+
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/offer/evaluate/security/CertificateNamesGeneratorTest.java:
```diff
@@ -0,0 +1,169 @@
+                networkNames);
+    }
+
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/offer/evaluate/security/SecretNameGeneratorTest.java:
```diff
@@ -0,0 +1,102 @@
+                secretNameGenerator.getCertificatePath());
+    }
+
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/offer/evaluate/security/TLSArtifactsGeneratorTest.java:
```diff
@@ -0,0 +1,168 @@
+                tlsArtifacts.getTrustStore().getCertificate("dcos-root"),
+                chain.get(chain.size() - 1));
+    }
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/offer/evaluate/security/TLSArtifactsPersisterTest.java:
```diff
@@ -0,0 +1,150 @@
+        verify(secretsClientMock, never()).delete(anyString());
+    }
+
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/test/java/com/mesosphere/sdk/scheduler/uninstall/TLSCleanupStepTest.java:
```diff
@@ -0,0 +1,113 @@
+        Assert.assertEquals(step.getStatus(), Status.COMPLETE);
+    }
+
+}
```
Missing trailing newline

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultService.java:
```diff
@@ -96,12 +103,30 @@ private void initService() {
+                                .build());
+                secretsClient = Optional.of(new DefaultSecretsClient(executor));
+            } catch (Exception e) {
+                LOGGER.error("Failed to create a secrets store client," +
```
When would this happen? What happens if we're on DC/OS Open?

mhrabovcin replied:
I've added following code:

```
            Optional<SecretsClient> secretsClient = Optional.empty();
            if (!TaskUtils.getTasksWithTLS(getServiceSpec()).isEmpty()) {
```

Having config validator and this condition the error we'd run TLS cleanup block only if any of tasks is configured with TLS. This should cover not displaying any message on DC/OS Open.

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/specification/DefaultService.java:
```diff
@@ -96,12 +103,30 @@ private void initService() {
+            } catch (Exception e) {
+                LOGGER.error("Failed to create a secrets store client," +
+                       "TLS artifacts possibly won't be cleaned up from secrets store");
+                LOGGER.error(String.valueOf(e));
```
This can be combined into the above as `LOGGER.error(String, Throwable)`

nickbp commented on frameworks/kafka/universe/config.json:
```diff
@@ -40,6 +40,11 @@
             "default": false
           },
+          "tls": {
+            "description": "Enable TLS",
```
Will this only work on EE clusters? If so, should probably mention that here.

mhrabovcin replied:
`Enable the TLS. This works only on Mesosphere DC/OS Enterprise installed in permissive or security modes.`

triclambert replied:
According to @keithchambers, please use 
> Enable TLS support (requires Enterprise DC/OS running in strict security mode)

mhrabovcin replied:
@triclambert should we say `... strict or permissive security mode`?

nickbp commented on testing/tls/keystore/config/integration-test.yml:
```diff
@@ -0,0 +1,35 @@
+    verifyHostname: true
+    trustStorePath: dropwizard.truststore
+    trustStorePassword: notsecure
+    trustSelfSignedCertificates: false
```
Unable to leave a comment on binary `local.keystore`/`local.truststore`. Are those supposed to be committed or should there be a .gitignore for them?

mhrabovcin replied:
I was planning to provide those as examples generated by scheduler to make possible local testing easier. Would you like to remove these items and add documentation how to create them?

nickbp replied:
If they're required in unit tests then I think it'd be fine to have them, otherwise I'd recommend omitting them

mhrabovcin replied:
I've removed files and added additional documentation about how to create files if necessary.

nickbp commented on testing/tls/keystore/README.md:
```diff
@@ -0,0 +1,89 @@
+# keystore-app
+
+The `keystore-app` is simplified version of `dropwizard-example` application
```
Move this under frameworks/helloworld/ since it's only used there?

nickbp commented on frameworks/helloworld/src/main/dist/examples/tls.yml:
```diff
@@ -0,0 +1,119 @@
+    uris:
+      - {{JAVA_URI}}
+      - {{BOOTSTRAP_URI}}
+      - https://s3-us-west-2.amazonaws.com/infinity-artifacts/testing/tls/keystore-app-{{KEYSTORE_APP_VERSION}}.zip
```
This should be built/uploaded as a part of the hello-world service

mhrabovcin commented on docs/pages/reference/yaml-reference.md:
```diff
@@ -352,7 +352,23 @@ This documentation effectively reflects the Java object tree under [RawServiceSp
 
-        The default visibility for the discovery information. May be `FRAMEWORK`, `CLUSTER`, or `EXTERNAL`. If unset this defaults to `CLUSTER`. See [Mesos documentation](http://mesos.apache.org/documentation/latest/app-framework-development-guide/) on service discovery for more information on these visibility values.
+        The default visibility for the discovery information. May be `FRAMEWORK`, `CLUSTER`, or `EXTERNAL`. If unset this defaults to `CLUSTER`. See [Mesos documentation]
+        (http://mesos.apache.org/documentation/latest/app-framework-development-guide/) on service discovery for more information on these visibility values.
```
This must be just how GH displays the URL. There is no newline.

nickbp replied:
I mean, at the moment it has:

```
See [Mesos documentation]
        (http://mesos.apache.org/documentation/latest/app-framework-development-guide/)
```

rather than:

```
See [Mesos documentation](http://mesos.apache.org/documentation/latest/app-framework-development-guide/)
```

mhrabovcin commented on frameworks/kafka/src/main/java/com/mesosphere/sdk/kafka/api/TopicResource.java:
```diff
@@ -105,11 +105,16 @@ public Response operationOnTopic(
                 switch (type) {
                     case "producer-test":
-                        int messageCount = Integer.parseInt(messages);
+                        messageCount = Integer.parseInt(messages);
```
Thanks, I've moved int parsing to method call and added braces.

mhrabovcin commented on sdk/scheduler/build.gradle:
```diff
@@ -102,6 +104,13 @@ dependencies {
     compile "org.glassfish.web:javax.el:${elVer}"
     compile 'com.mesosphere:mesos-http-adapter:0.4.1'
+    compile("com.auth0:java-jwt:${jwtVer}") {
+        // JWT package comes with jackson version 2.8.4 which doesn't work properly with
```
I've actually tried to bump the library to the `2.8.4` version. The new version was a reason why loopback serialization wasn't working and hello-world placement integration tests were failing.. I've tried to debug it but didn't find a reason in a day so I've decided to keep existing version and possibly tackle the Jackson upgrade in a separate PR.

nickbp replied:
Interesting, yeah was just curious, no need to bump if it's causing problems.

mhrabovcin commented on frameworks/kafka/universe/marathon.json.mustache:
```diff
@@ -48,6 +48,7 @@
     "BROKER_DISK_PATH": "{{brokers.disk_path}}",
     "BROKER_JAVA_HEAP": "{{brokers.heap.size}}",
     "BROKER_PORT": "{{brokers.port}}",
+    "BROKER_PORT_TLS": "{{brokers.port_tls}}",
```
Moved.

mhrabovcin commented on frameworks/kafka/universe/config.json:
```diff
@@ -111,6 +116,11 @@
             "default": 0
           },
+          "port_tls": {
+            "description": "Port for broker to listen on for TLS connections",
```
Added.

mhrabovcin commented on sdk/bootstrap/secret_decoder.go:
```diff
@@ -0,0 +1,75 @@
+		dstData := make([]byte, base64.StdEncoding.EncodedLen(len(srcData)))
+		base64.StdEncoding.Decode(dstData, srcData)
+		// Remove any trailing zero bytes
+		dstData = bytes.Trim(dstData, "\x00")
```
We're pre allocating bytes array before running the decode function:

```
dstData := make([]byte, base64.StdEncoding.EncodedLen(len(srcData)))
```
The `EncodedLen` function returns maximum possible length of encoded data rather than actual length that will be determined during the decoding. I've added `TrimRight` but I expect this function to be removed once we get support for base64 decoding from the mesos secrets module.

nickbp replied:
:+1:

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/CertificateAuthorityClient.java:
```diff
@@ -0,0 +1,37 @@
+     * @throws CertificateException
+     * @throws CAException
+     */
+    Collection<X509Certificate> chainWithRootCert(
```
👍

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public CachedTokenProvider(TokenProvider provider) {
+        this(provider, 30);
```
Great suggestion. Added.

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/ca/DefaultCAClient.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    @Override
+    public X509Certificate sign(byte[] csr) throws Exception {
```
There is also possibility of `java.security.cert.CertificateException` when PEM string is decoded. I had originally 2 distinct exceptions and moved it to common single Exception.

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/http/URLHelper.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * URLHelper for easier URL building.
+ */
+public class URLHelper {
```
What private constructor is required? (Renamed)

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java:
```diff
@@ -0,0 +1,355 @@
+     */
+    public static class Builder {
+
+        @Valid
```
I've added validation rules and `ValidationUtils` call to the `build` method.

nickbp commented on sdk/bootstrap/secret_decoder.go:
```diff
@@ -0,0 +1,75 @@
+	}
+
+	truststoreMatches, err := filepath.Glob(
+		d.path + "/**.truststore.base64")
```
Another double asterisk here

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/auth/CachedTokenProvider.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    public CachedTokenProvider(TokenProvider provider) {
+        this(provider, Duration.ofSeconds(30));
```
Could remove this version of the constructor, now that the default is being handled upstream in SchedulerFlags

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/ca/PEMUtils.java:
```diff
@@ -0,0 +1,35 @@
+ * A {@link PEMUtils} allows to convert X509 certificates from and to PEM format.
+ */
+public class PEMUtils {
+
```
Should have an empty constructor:

```
private PEMUtils() {
    // do not instantiate
}
```

mhrabovcin replied:
Added.

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/http/URLUtils.java:
```diff
@@ -0,0 +1,42 @@
+ */
+public class URLUtils {
+
+    private static final Logger logger = LoggerFactory.getLogger(URLUtils.class);
```
Should have an empty constructor:

```
private URLUtils() {
    // do not instantiate
}
```

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/dcos/ca/DefaultCAClient.java:
```diff
@@ -0,0 +1,159 @@
+     * @throws CAException
+     */
+    protected void handleResponseStatusLine(StatusLine statusLine, int okCode) throws CAException {
+        if (statusLine.getStatusCode() == okCode) {
```
Would it make sense to just throw if `statusLine.getStatusCode() != okCode`? The presence of okCode implies that we know exactly what we expect here.

mhrabovcin replied:
I've implemented suggested approach with `statusLine.getStatusCode() != okCode`.

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/OfferEvaluator.java:
```diff
@@ -411,13 +437,32 @@ private static void logOutcome(StringBuilder stringBuilder, EvaluationOutcome ou
                     .forEach(resource -> evaluationStages.add(new UnreserveEvaluationStage(resource)));
             evaluationStages.addAll(taskResourceMapper.getEvaluationStages());
 
+            if (taskSpec.getTransportEncryption().size() > 0) {
```
May as well use `!isEmpty()` to align with the similar check at L332

nickbp commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/config/validate/TLSRequiresServiceAccount.java:
```diff
@@ -0,0 +1,48 @@
+        if (tasksWithTLSCount > 0) {
+            try {
+                if (flags.getServiceAccountUid().equals("")) {
+                    throw new Exception("Missing service account");
```
Just return a ConfigValidationError in this case?

mhrabovcin commented on sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/uninstall/UninstallScheduler.java:
```diff
@@ -87,14 +108,26 @@ private Plan getPlan() {
+        phases.add(resourcePhase);
+
+        if (secretsClient.isPresent()) {
+            Step tlsCleanupStep = new TLSCleanupStep(
```
1. It may be a valid request to keep the TLS artifacts. Would you like to implement another scheduler flag which would control such a behavior?

2. In case of any failure the TLS certificates would be reused (if artifacts are complete in secret store) or overwritten when existing values couldn't be used. In other words, failed uninstall wouldn't prevent reinstalling the service.

mhrabovcin commented:
@nickbp @benclarkwood I've removed `kafka` changes, squashed all commits and we now have only 1 integration tests failure:

https://teamcity.mesosphere.io/viewLog.html?buildId=731824&buildTypeId=DcosIo_DcosCommons_Hdfs_BuildAndTestPr&guest=1

The `test_shakedown.test_modify_app_config` is reported in the UI as flaky.

I believe this is ready to be merged.

benclarkwood commented:
Yes, HDFS is very flaky. I will attempt to get a non-flake today.

@mhrabovcin, please create a ticket to update the logic of the TLSEvaluationStage creation to not log an error when no TLS is requested. It will confuse users and lead to unnecessary support traffic :).

mhrabovcin commented:
Ticket created - https://jira.mesosphere.com/browse/DCOS-17685

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+Standard [PEM encoded](https://en.wikipedia.org/wiki/Privacy-enhanced_Electronic_Mail) certificate, private key and root CA certificate. Based on a file extension name it is possible to tell which artifact is in the file.
+
+A certificate file with **.crt** extension contains an end-entity certificate with possible chain of certificates leading to the root CA certificate. The root CA certificate is not part of end-entity certificate file.
```
"with an optional chain of certificates"?

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+A certificate file with **.crt** extension contains an end-entity certificate with possible chain of certificates leading to the root CA certificate. The root CA certificate is not part of end-entity certificate file.
+
+A **.key** file contains PEM encoded private key.
```
contains the

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+A **.key** file contains PEM encoded private key.
+
+A file with **.crt** extension contains root CA certificate without certificate chain.
```
the root CA certificate without the certificate chain

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+#### Java Keystore
+
+The [Java Keystore (JKS)](https://en.wikipedia.org/wiki/Keystore) is a repository of various certificates and private keys. When private key and certificate is requested a task will get 2 JKS format files.
```
When a private key and certificate is requested, a task...

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+The [Java Keystore (JKS)](https://en.wikipedia.org/wiki/Keystore) is a repository of various certificates and private keys. When private key and certificate is requested a task will get 2 JKS format files.
+
+A **.keystore** is a JKS file that contains a private key with end-entity certificate and complete certificate chain including the root CA certificate. Certificate is stored under alias name **`default`**. The keystore and key are protected by single **`notsecure`** password.
```
an end-entity certificate and a complete certificate chain, including the root CA certificate

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+A **.keystore** is a JKS file that contains a private key with end-entity certificate and complete certificate chain including the root CA certificate. Certificate is stored under alias name **`default`**. The keystore and key are protected by single **`notsecure`** password.
+
+A **.truststore** is a JKS file that contains DC/OS root CA certificate stored as a trust certificate. The certificte is stored under alias **`dcos-root`**. The keystore is protected by **`notsecure`** password.
```
the DC/OS root..

The certificate is stored under the alias...

The keystore is protected by the password **`notsecure`**.

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+A **.truststore** is a JKS file that contains DC/OS root CA certificate stored as a trust certificate. The certificte is stored under alias **`dcos-root`**. The keystore is protected by **`notsecure`** password.
+
+The password **`notsecure`** that is protecting both JKS files (containing an end-entity certificate with private key and root CA certificate) is selected because most of Java tools and libraries validate for some password. It is not to meant to provide any additional protection. Security of both files is achieved by using DC/OS secrets store with file-based in-memory secrets. No other schedulers ...
```
that protects both

...has been selected because most Java tools and libraries require a password.

Both files are secured using the DC/OS secrets store...

No other schedulers or tasks can access...

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+The password **`notsecure`** that is protecting both JKS files (containing an end-entity certificate with private key and root CA certificate) is selected because most of Java tools and libraries validate for some password. It is not to meant to provide any additional protection. Security of both files is achieved by using DC/OS secrets store with file-based in-memory secrets. No other schedulers ...
+
+**Imporant note:** The DC/OS 1.10 secrets store doesn't provide support for storing binary data. To overcome this limitation the `keystore` and `truststore` files are encoded using Base64 encoding when stored to secrets store. Files are mounted into a task container with additional `base64` file name suffix - `*.keystore.base64` and `*.truststore.base64`. There is a `bootstrap` utility helper tha ...
```
To overcome this limitation, the `keystore` and `truststore` files are Base64-encoded before they are added to the secrets store.

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+To enable TLS support a `Mesosphere DC/OS Enterprise` cluster must be installed in `permissive` or `strict` security mode.
+
+A scheduler must run with `service account` that has permissions to access:
```
with a service account that has permission to access:

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+To enable TLS support a `Mesosphere DC/OS Enterprise` cluster must be installed in `permissive` or `strict` security mode.
+
+A scheduler must run with `service account` that has permissions to access:
+
```
periods after list items.

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+A scheduler must run with `service account` that has permissions to access:
+
+- `DC/OS CA` - requires `full` permission to [`dcos:adminrouter:ops:ca:rw`](https://docs.mesosphere.com/1.9/security/perms-reference/#admin-router) resource
```
permission to the

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+- `DC/OS CA` - requires `full` permission to [`dcos:adminrouter:ops:ca:rw`](https://docs.mesosphere.com/1.9/security/perms-reference/#admin-router) resource
+
+- Secrets store - requires `full` permission [`dcos:secrets:default:<service-name>/`](https://docs.mesosphere.com/1.9/security/perms-reference/#secrets) and `full` permission on `dcos:secrets:list:default:<service-name>`
```
permission to

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+- Secrets store - requires `full` permission [`dcos:secrets:default:<service-name>/`](https://docs.mesosphere.com/1.9/security/perms-reference/#secrets) and `full` permission on `dcos:secrets:list:default:<service-name>`
+
+  This is a minimal permission that would allow scheduler to store secrets in service namespace. However with only this permission a scheduler wouldn't be able to clean up created secrets when service is uninstalled. This is a known limitation and will be addressed in `DC/OS` version `1.11`.
```
allow the scheduler to store secrets in the service namespace. However, with only this permission, a scheduler wouldn't be able to clean up created secrets when the service is uninstalled.

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+  This is a minimal permission that would allow scheduler to store secrets in service namespace. However with only this permission a scheduler wouldn't be able to clean up created secrets when service is uninstalled. This is a known limitation and will be addressed in `DC/OS` version `1.11`.
+
+  Alternatively it is possible to assign service account broad `full` permission on `dcos:superuser` resource. This is a security risk so it is not adviced to run any scheduler with such a permission.
```
Alternatively, you can assign a service account broad

This is a security risk, so we do not advise you to run a scheduler with this permission.

sascala commented on docs/pages/developer-guide.md:
```diff
@@ -936,6 +935,174 @@ In the example above, a server task can be accessed through the address:
+
+  Alternatively it is possible to assign service account broad `full` permission on `dcos:superuser` resource. This is a security risk so it is not adviced to run any scheduler with such a permission.
+
+  Operator can manually clean up secrets via DC/OS UI or CLI before uninstalling service that uses TLS. If operator would like to use `dcos:superuser` permission for automated uninstall it would be possible to assign such permission just before uninstalling package and removing the permission from account after the service has been uninstalled.
```
The operate can clean up secrets manually via the DC/OS GUI or CLI before uninstalling a service... 

If the operator would like to use the `dcos:superuser` permission for automated uninstall, it would be possible to assign this permission just before uninstalling the package and remove it from account after the service has been uninstalled.

sascala commented on docs/pages/reference/yaml-reference.md:
```diff
@@ -354,6 +354,21 @@ This documentation effectively reflects the Java object tree under [RawServiceSp
 
         The default visibility for the discovery information. May be `FRAMEWORK`, `CLUSTER`, or `EXTERNAL`. If unset this defaults to `CLUSTER`. See [Mesos documentation](http://mesos.apache.org/documentation/latest/app-framework-development-guide/) on service discovery for more information on these visibility values.
 
+    * `transport-encryption`
```
If unset, this defaults...

sascala commented on docs/pages/reference/yaml-reference.md:
```diff
@@ -354,6 +354,21 @@ This documentation effectively reflects the Java object tree under [RawServiceSp
+    * `transport-encryption`
+
+      A task may optionally ask for a X.509 TLS certificate with private key and CA certificate bundle. A certificate can be used by service to enable secure communication.
+
```
by the service

sascala commented on docs/pages/reference/yaml-reference.md:
```diff
@@ -354,6 +354,21 @@ This documentation effectively reflects the Java object tree under [RawServiceSp
+
+      * `name`
+
+        A name of files representing the TLS artifacts in the task sandbox directory. For example a `name: nginx` with `type: TLS` will result in `$MESOS_SANDBOX/nginx.crt`, `$MESOS_SANDBOX/nginx.key` and `$MESOS_SANDBOX/nginx.ca` files.
```
A name for the files represent the TLS artifacts...

sascala commented on docs/pages/reference/yaml-reference.md:
```diff
@@ -354,6 +354,21 @@ This documentation effectively reflects the Java object tree under [RawServiceSp
+
+        A type or format of delivered TLS artifacts.
+        This can be set either to `TLS` for PEM encoded private key file, certificate and CA bundle or `KEYSTORE` for certificate and private key to be delivered in a separate keystore file and CA bundle in other truststore file.
+
```
comma after "bundle"

sascala commented on docs/pages/reference/yaml-reference.md:
```diff
@@ -354,6 +354,21 @@ This documentation effectively reflects the Java object tree under [RawServiceSp
+        A type or format of delivered TLS artifacts.
+        This can be set either to `TLS` for PEM encoded private key file, certificate and CA bundle or `KEYSTORE` for certificate and private key to be delivered in a separate keystore file and CA bundle in other truststore file.
+
+      For detailed information see the [SDK Developer Guide](developer-guide.html#tls).
```
coma after "information"

elezar commented:
Closing to see if this retriggers the tests.
